charts/partials/templates/_proxy.tpl

```diff
@@ -182,22 +190,28 @@ securityContext:
   seccompProfile:
     type: RuntimeDefault
 terminationMessagePolicy: FallbackToLogsOnError
-{{- if or (.Values.proxy.await) (.Values.proxy.waitBeforeExitSeconds) }}
+{{- if or .Values.proxy.await .Values.proxy.waitBeforeExitSeconds .Values.proxy.nativeSidecar }}
 lifecycle:
-{{- if .Values.proxy.await }}
+{{- if and .Values.proxy.await (not .Values.proxy.nativeSidecar) }}
   postStart:
     exec:
       command:
         - /usr/lib/linkerd/linkerd-await
         - --timeout=2m
         - --port={{.Values.proxy.ports.admin}}
 {{- end }}
-{{- if .Values.proxy.waitBeforeExitSeconds }}
+{{- if or .Values.proxy.waitBeforeExitSeconds .Values.proxy.nativeSidecar }}
   preStop:
     exec:
       command:
+        {{- if .Values.proxy.nativeSidecar }}
+        - /usr/lib/linkerd/linkerd-await
+        - --timeout=1s
+        - --port={{.Values.proxy.ports.admin}}
+        - --shutdown
+        {{- end }}
         - /bin/sleep
-        - {{.Values.proxy.waitBeforeExitSeconds | quote}}
+        - {{.Values.proxy.waitBeforeExitSeconds | default 0 | quote}}
```

Comment:
Testing this in k8s `v1.29.0-alpha.3` I realized we need to get rid of this hook. According to the KEP:
> 1. Sidecar containers that have a PreStop hook will be notified when the Pod has begun terminating by executing the PreStop hook. This happens at the same time as regular containers, and begins the Pod's termination grace period countdown.
> 2. Once the last primary container terminates, the last started sidecar container is notified by sending a SIGTERM signal.

With this hook, If `waitBeforeExitSeconds` is zero, then the proxy will be terminated as soon as the pod termination starts, which is not what we want. If we remove this hook, then the proxy will terminate only after the primary container has terminated.

Reply:
So if I'm reading it right `POST /shutdown` and SIGTERM do the same thing to the proxy?

https://github.com/linkerd/linkerd2-proxy/blob/main/linkerd2-proxy/src/main.rs#L126-L141

I spent a lot of time staring at this doc: https://linkerd.io/2.14/tasks/graceful-shutdown/#configuration-options-for-graceful-shutdown

My thought process is that this shutdown hook will start the grace period similar to what sigterm used to do to bias new connections away from the pod as it is shutting down.

Reply:
You're correct about `/shutdown` and SIGTERM having the same effect on the proxy.
I can see there could be room here for an additional touch point to start the draining process. There's however an ongoing debate (see #10379) whether that draining is desirable at all. It seems the new native sidecar workflow without proxy connection draining would address the most common cases. Perhaps there's value on adding more knobs to address the special cases, which could be explored as a separate effort.